### PR TITLE
Add checks to schedulers

### PR DESCRIPTION
**Pull request by b-chu**
# What does this PR do?

Changes the error checking in scheduler to check
1. if max_duration is greater than t_max
2. if t_warmup and t_max are different units

Previously it only checked if t_warmup and max_duration were different units which is incorrect because a valid schedule can have t_warmup and max_duration as different units.

# What issue(s) does this change relate to?

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

File: composer/optim/scheduler.py
Comment (snarayan21) [inline composer/optim/scheduler.py:553]: why not warning here? Also, i think the message could be more explicit. Instead of "make sure that your LR schedule is defined at all points", we probably want to say something that tells the user that max_duration should never exceed t_max, or their LR schedule will be invalid.

File: composer/optim/scheduler.py
Comment (snarayan21) [inline composer/optim/scheduler.py:568]:
```suggestion
    _convert_time should be called on both `t_warmup` and `t_max` before this function is called. As a a result, t_warmup and t_max will not
```

File: tests/optim/test_scheduler.py
Comment (snarayan21) [inline tests/optim/test_scheduler.py:443]: just for my understanding, what does `nullcontext()` help us do

File: tests/optim/test_scheduler.py
Comment (snarayan21) [inline tests/optim/test_scheduler.py:466]: can you use the `TimeUnit` enums here instead
https://github.com/mosaicml/composer/blob/2f11230a17bd35eb4f6a5784a18e43c648a56b2f/composer/core/time.py#L32

**Review by snarayan21 — COMMENTED** [review]
some minor comments!

File: composer/optim/scheduler.py
Comment (b-chu) [reply composer/optim/scheduler.py:553]: Every optimizer step will print to the logs and there wouldn't be an easy way to filter it out.

"max_duration should never exceed t_max" Some schedulers have a defined schedule for duration above t_max

File: tests/optim/test_scheduler.py
Comment (b-chu) [reply tests/optim/test_scheduler.py:443]: It basically says that there won't be an error for that test case

File: tests/optim/test_scheduler.py
Comment (b-chu) [reply tests/optim/test_scheduler.py:466]: I'll leave this to a separate PR since pytest params should also use TimeUnit enums if this change is made

File: composer/optim/scheduler.py
Comment (snarayan21) [reply composer/optim/scheduler.py:553]: wait really? could you link which ones -- pretty sure that every scheduler that takes in a `t_max` is only supposed to be defined up to that

There are other schedulers with just warmup that don't take t_max, yeah

File: composer/optim/scheduler.py
Comment (snarayan21) [inline composer/optim/scheduler.py:524]: shouldn't we raise error if `frac_of_total > 1.0`? As in, if current time exceeds t_max? This is the slow failing behavior we talked about right

**Review by snarayan21 — APPROVED** [review]
lgtm, would appreciate clarifiation on questions before merging though.

File: composer/optim/scheduler.py
Comment (b-chu) [reply composer/optim/scheduler.py:553]: Linked some examples below

File: composer/optim/scheduler.py
Comment (b-chu) [reply composer/optim/scheduler.py:524]: Seems like most schedulers account for this behavior already, but it was missing from this one so adding it to make all schedulers consistent

https://github.com/mosaicml/composer/blob/2f11230a17bd35eb4f6a5784a18e43c648a56b2f/composer/optim/scheduler.py#L350
https://github.com/mosaicml/composer/blob/2f11230a17bd35eb4f6a5784a18e43c648a56b2f/composer/optim/scheduler.py#L394
https://github.com/mosaicml/composer/blob/2f11230a17bd35eb4f6a5784a18e43c648a56b2f/composer/optim/scheduler.py#L747
https://github.com/mosaicml/composer/blob/2f11230a17bd35eb4f6a5784a18e43c648a56b2f/composer/optim/scheduler.py#L822
https://github.com/mosaicml/composer/blob/2f11230a17bd35eb4f6a5784a18e43c648a56b2f/composer/optim/scheduler.py#L899

**Pull request merged**
